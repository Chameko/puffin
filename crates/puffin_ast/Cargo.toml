[package]
name = "puffin_ast"
version = "0.0.1"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]

[dependencies]
rowan = "0.15"