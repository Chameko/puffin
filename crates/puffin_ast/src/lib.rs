/// Describes all the syntax kinds in puffin
pub mod syntax_kind;
pub mod ast;
pub mod green;

use puffin_source::TextSlice;
use rowan::TextRange;
pub use syntax_kind::SyntaxKind;
pub use green::{SyntaxNode, SyntaxToken, SyntaxNodeChildren};
pub use ast::{AstPtr, SyntaxNodePtr, AstMap};

/// Converts between a [TextSlice] and a [TextRange]
pub fn text_range(slice: TextSlice) -> TextRange {
    TextRange::new((*slice.start()).into(), (*slice.end() + 1).into())
}
