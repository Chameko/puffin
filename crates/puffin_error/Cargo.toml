[package]
name = "puffin_error"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
puffin_hir = { path = "../puffin_hir/" }
puffin_vfs = { path = "../puffin_vfs/" }
thiserror = "1.0.43"
colored = "2.0.4"
