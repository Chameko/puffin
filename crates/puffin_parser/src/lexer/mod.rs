pub mod lexer;
pub mod tokenstream;
pub use lexer::Lexer;
use puffin_vfs::FileID;
pub use tokenstream::TokenStream;
pub use puffin_hir::source::SourceDatabase;

#[salsa::query_group(LexerStorage)]
pub trait LexerDatabase : SourceDatabase {
    /// Get the scanned tokens
    fn scan(&self, file: FileID) -> TokenStream;
}

/// Scan the source into tokens
fn scan(db: &dyn LexerDatabase, file: FileID) -> TokenStream {
    let src_tree = db.source_tree();
    let src = src_tree.find_source(file).expect("file should be in source tree");
    let lexer = lexer::Lexer::new(src.get_text());
    let tokens = lexer.start_scan();
    TokenStream::new(tokens)
}
