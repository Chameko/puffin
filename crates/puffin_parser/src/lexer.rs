use puffin_ast::SyntaxKind;
use std::{iter::Peekable, str::Chars, ops::RangeInclusive};

/// A token used in the lexer and parser
#[derive(Debug, Clone)]
pub struct Token {
    /// The type of the token
    pub ty: SyntaxKind,
    /// The column on the line the token is on. The indexing starts from 0
    pub col: RangeInclusive<usize>,
    /// The line the token is on. The indexing starts from 1
    pub line: usize,
}

impl Token {
    /// Creates a new token
    pub fn new(ty: SyntaxKind, col: RangeInclusive<usize>, line: usize) -> Self {
        Self {
            ty,
            col,
            line,
        }
    }

    /// Extracts the tokens text from the provided source, split by lines.
    pub fn get_text<'a>(&self, src: &Vec<&'a str>) -> &'a str {
        src[self.line - 1].get(self.col.clone()).expect("Token out of range")
    }
}

/// The lexer responsible for taking a raw file as a string and converting it to a flat array of [`Token`]
pub struct Lexer<'a> {
    /// The source the lexer is parsing
    src: Peekable<Chars<'a>>,
    /// Charaters that may form a keyword or identifier are stored here
    working: String,
    /// The current line
    line: usize,
    /// The current column
    col: usize,
    /// If the next character is on a newline
    next_line: bool,
}

impl<'a> Lexer<'a> {
    /// Next character
    #[inline]
    fn next(&mut self) -> Option<char> {
        let nxt = self.src.next();
        // Add a line if neccessary
        if self.next_line {
            self.line += 1;
            self.col = 0;
            self.next_line = false;
        }
        // Prepare for next line
        if let Some('\n') = nxt {
            self.next_line = true;
        }
        nxt
    }

    /// Peek one character ahead
    #[inline]
    fn peek(&mut self) -> Option<char> {
        self.src.peek().map(|c| *c)
    }

    ///  Creates a new lexer
    pub fn new(src: &'a str) -> Self {
        Self {
            src: src.chars().peekable(),
            working: String::new(),
            next_line: false,
            line: 1,
            col: 0,
        }
    }

    /// Parses a number
    fn parse_number(s: &str) -> SyntaxKind {
        let mut ty = SyntaxKind::INT;
        for char in s.chars() {
            if !char.is_numeric() {
                if char == '.' {
                    ty = SyntaxKind::FLOAT;
                } else {
                    // Cannot have letters in a number
                    return ty;
                }
            }
        }
        ty
    }

    // Scans the working string and creates an identifier or a number
    fn scan_working(&mut self) -> Option<Token> {
        let ret = match self.working.as_str() {
            "and" => Some(SyntaxKind::KW_AND),
            "or" => Some(SyntaxKind::KW_OR),
            "print" => Some(SyntaxKind::KW_PRINT),
            "let" => Some(SyntaxKind::KW_LET),
            s => {
                if let Some(char) = s.chars().next() {
                    if char.is_numeric() {
                        Some(Self::parse_number(s))
                    } else {
                        Some(SyntaxKind::IDENT)
                    }
                } else {
                    None
                }
            }
        };
        let working_clone = std::mem::take(&mut self.working) ;
        if let Some(ty) = ret {
            let ret = Some(Token::new(ty, self.col..=(self.col + working_clone.len() - 1), self.line));
            self.col += working_clone.len();
            ret
        } else {
            None
        }
    }

    /// Starts the scan of the file
    pub fn start_scan(mut self) -> Vec<Token> {
        let mut tokens = vec![];
        self.scan(&mut tokens);
        tokens
    }

    /// Used to record a symbol. Note: parses the working string before recording the symbol
    fn symbol(&mut self, ty: SyntaxKind, string: &str, tokens: &mut Vec<Token>) {
        if let Some(tk) = self.scan_working() {
            tokens.push(tk);
        }
        tokens.push(Token::new(ty, self.col..=(self.col + string.len() - 1), self.line));
        self.col += string.len();
    }

    /// Scans the string into a flat array of [`SyntaxKind`] and [`String`]
    fn scan(&mut self, tokens: &mut Vec<Token>) {
        if let Some(char) = self.next() {
            match char {
                '&' => {
                    if let Some('&') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::AMPAMP, "&&", tokens);
                    } else {
                        self.symbol( SyntaxKind::AMP, "&", tokens);
                    }
                }
                '|' => {
                    if let Some('|') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::PIPEPIPE, "||", tokens);
                    } else {
                        self.symbol( SyntaxKind::PIPE, "|", tokens);
                    }
                }
                '=' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::EQEQ, "==", tokens);
                    } else {
                        self.symbol( SyntaxKind::EQ, "=", tokens)
                    }
                }
                '>' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::GTEQ, ">=", tokens);
                    } else {
                        self.symbol( SyntaxKind::GT, ">", tokens);
                    }
                }
                '<' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::LTEQ, "<=", tokens);
                    } else {
                        self.symbol( SyntaxKind::LT, "<", tokens);
                    }
                }
                '.' => {
                    if let Some(c) = self.peek() {
                        // If the . is part of a number we push it
                        if c.is_numeric() {
                            self.working.push('.')
                        } else {
                            self.symbol( SyntaxKind::DOT, ".", tokens);
                        }
                    } else {
                        self.symbol( SyntaxKind::DOT, ".", tokens)
                    }
                }
                '/' => {
                    if let Some('/') = self.peek() {
                        self.next();
                        let mut peek = self.peek();
                        let mut comment = String::new();
                        while peek != Some('\n') && peek != None {
                            comment.push(self.next().unwrap());
                            peek = self.peek();
                        }
                        self.symbol( SyntaxKind::COMMENT, "//", tokens);
                    } else {
                        self.symbol( SyntaxKind::SLASH, "/", tokens);
                    }
                }
                ' ' => {
                    let mut length = 1;
                    let mut col = self.col;
                    while let Some(' ') = self.peek() {
                         length += 1;
                         self.next();
                    }
                    if let Some(tk) = self.scan_working() {
                        // Update the column as we may have scanned a new token
                        col = *tk.col.end() + 1;
                        tokens.push(tk);
                    }
                    tokens.push(Token::new(SyntaxKind::WHITESPACE, col..=(col + length - 1), self.line));
                    self.col += length;
                }
                '+' => self.symbol( SyntaxKind::PLUS, "+", tokens),
                '-' => self.symbol( SyntaxKind::MINUS, "-", tokens),
                '*' => self.symbol( SyntaxKind::STAR, "*", tokens),
                '(' => self.symbol( SyntaxKind::L_PAREN, "(", tokens),
                ')' => self.symbol( SyntaxKind::R_PAREN, ")", tokens),
                '{' => self.symbol(SyntaxKind::L_BRACE, "{", tokens),
                '}' => self.symbol(SyntaxKind::R_PAREN, "}", tokens),
                '!' => self.symbol( SyntaxKind::EXCLAMATION, "!", tokens),
                '\n' => self.symbol( SyntaxKind::NL, "\n", tokens),
                c => {
                    if c.is_alphanumeric() {
                        self.working.push(c)
                    } else {
                        self.symbol( SyntaxKind::ERROR, &c.to_string(), tokens)
                    }
                }
            }
            self.scan(tokens);
        } else {
            if let Some(tk) = self.scan_working() {
                tokens.push(tk);
            }
        }
    }
}

#[cfg(test)]
mod lexer_test {
    use super::Token;

    fn test_base(input: &str) -> String {
        let lexer = super::Lexer::new(input);
        let tokens = lexer.start_scan();
        let input = input.split_inclusive('\n').collect::<Vec<&str>>();
        tokens.into_iter().map(|t| {
            input[t.line - 1].get(t.col).unwrap()
        }).collect::<String>()
    }

    #[test]
    fn symbol_kinds() {
        let output = test_base("& | + - * / > < = ( ) ! . \n == >= <= && || { }");
        insta::assert_snapshot!(output);
    }

    #[test]
    fn keywords_and_ident() {
        let input = "and sometoiurng or skngeing";
        let lexer = super::Lexer::new(input);
        let tokens = lexer.start_scan();
        let input = input.split('\n').collect::<Vec<&str>>();
        let output = tokens
            .into_iter()
            .map(|p| {
                match p {
                Token {ty: super::SyntaxKind::KW_AND, ..} => format!("|and:{}|", input[p.line - 1].get(p.col).unwrap()),
                Token {ty: super::SyntaxKind::KW_OR, ..} => format!("|or:{}|", input[p.line - 1].get(p.col).unwrap()),
                Token {ty: super::SyntaxKind::WHITESPACE, ..} => format!("|whitespace:{}|", input[p.line - 1].get(p.col).unwrap()),
                Token {ty: super::SyntaxKind::IDENT, ..} => format!("|ident:{}|", input[p.line - 1].get(p.col).unwrap()),
                Token {ty: super::SyntaxKind::SOURCE_FILE, ..}  => format!("|src:{}|", input[p.line - 1].get(p.col).unwrap()),
                _ => format!("error:{}", input[p.line].get(p.col).unwrap()),
            }})
            .collect::<String>();
        insta::assert_snapshot!(output);
    }

    #[test]
    fn numbers() {
        let input = "1.23 42";
        let lexer = super::Lexer::new(input);
        let input = input.split('\n').collect::<Vec<&str>>();
        let tokens = lexer.start_scan();
        let output = tokens
            .into_iter()
            .map(|p| match p {
                Token {ty: super::SyntaxKind::FLOAT, .. } => format!("|float:{}|", input[p.line - 1].get(p.col).unwrap()),
                Token {ty: super::SyntaxKind::INT, .. } => format!("|int:{}|", input[p.line - 1].get(p.col).unwrap()),
                Token {ty: super::SyntaxKind::WHITESPACE, .. } => format!("|whitespace:{}|", input[p.line - 1].get(p.col).unwrap()),
                Token {ty: super::SyntaxKind::SOURCE_FILE, .. }  => format!("|src:{}|", input[p.line - 1].get(p.col).unwrap()),
                _ => format!("error:{}", input[p.line].get(p.col).unwrap()),
            })
            .collect::<String>();
        insta::assert_snapshot!(output);
    }
}
