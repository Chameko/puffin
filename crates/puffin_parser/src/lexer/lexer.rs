use puffin_ast::SyntaxKind;
use std::{iter::Peekable, str::Chars};
use super::tokenstream::Token;

/// The lexer responsible for taking a raw file as a string and converting it to a flat array of [`Token`]
pub struct Lexer<'a> {
    /// The source the lexer is parsing
    src: Peekable<Chars<'a>>,
    /// Charaters that may form a keyword or identifier are stored here
    working: String,
    /// The current column
    col: u32,
}

impl<'a> Lexer<'a> {
    /// Next character
    #[inline]
    fn next(&mut self) -> Option<char> {
        self.src.next()
    }

    /// Peek one character ahead
    #[inline]
    fn peek(&mut self) -> Option<char> {
        self.src.peek().map(|c| *c)
    }

    ///  Creates a new lexer
    pub fn new(src: &'a str) -> Self {
        Self {
            src: src.chars().peekable(),
            working: String::new(),
            col: 0,
        }
    }

    /// Parses a number
    fn parse_number(s: &str) -> SyntaxKind {
        let mut ty = SyntaxKind::INT;
        for char in s.chars() {
            if !char.is_numeric() {
                if char == '.' {
                    ty = SyntaxKind::FLOAT;
                } else {
                    // Cannot have letters in a number
                    return ty;
                }
            }
        }
        ty
    }

    // Scans the working string and creates an identifier or a number
    fn scan_working(&mut self) -> Option<Token> {
        let ret = match self.working.as_str() {
            "and" => Some(SyntaxKind::KW_AND),
            "or" => Some(SyntaxKind::KW_OR),
            "print" => Some(SyntaxKind::KW_PRINT),
            "let" => Some(SyntaxKind::KW_LET),
            "fun" => Some(SyntaxKind::KW_FUN),
            "int" => Some(SyntaxKind::KW_INT),
            "float" => Some(SyntaxKind::KW_FLOAT),
            "bool" => Some(SyntaxKind::KW_BOOL),
            "string" => Some(SyntaxKind::KW_STRING),
            s => {
                if let Some(char) = s.chars().next() {
                    if char.is_numeric() {
                        Some(Self::parse_number(s))
                    } else {
                        Some(SyntaxKind::IDENT)
                    }
                } else {
                    None
                }
            }
        };
        let working_clone = std::mem::take(&mut self.working);
        if let Some(ty) = ret {
            let ret = Some(Token::new(ty, self.col..=(self.col + working_clone.len() as u32 - 1)));
            self.col += working_clone.len() as u32;
            ret
        } else {
            None
        }
    }

    /// Starts the scan of the file
    pub fn start_scan(mut self) -> Vec<Token> {
        let mut tokens = vec![];
        self.scan(&mut tokens);
        tokens
    }

    /// Used to record a symbol. Note: parses the working string before recording the symbol
    fn symbol(&mut self, ty: SyntaxKind, string: &str, tokens: &mut Vec<Token>) {
        if let Some(tk) = self.scan_working() {
            tokens.push(tk);
        }
        tokens.push(Token::new(ty, self.col..=(self.col + string.len() as u32 - 1)));
        self.col += string.len() as u32;
    }

    /// Scans the string into a flat array of [`SyntaxKind`] and [`String`]
    fn scan(&mut self, tokens: &mut Vec<Token>) {
        if let Some(char) = self.next() {
            match char {
                '&' => {
                    if let Some('&') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::AMPAMP, "&&", tokens);
                    } else {
                        self.symbol( SyntaxKind::AMP, "&", tokens);
                    }
                }
                '|' => {
                    if let Some('|') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::PIPEPIPE, "||", tokens);
                    } else {
                        self.symbol( SyntaxKind::PIPE, "|", tokens);
                    }
                }
                '=' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::EQEQ, "==", tokens);
                    } else {
                        self.symbol( SyntaxKind::EQ, "=", tokens)
                    }
                }
                '>' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::GTEQ, ">=", tokens);
                    } else {
                        self.symbol( SyntaxKind::GT, ">", tokens);
                    }
                }
                '<' => {
                    if let Some('=') = self.peek() {
                        self.next();
                        self.symbol( SyntaxKind::LTEQ, "<=", tokens);
                    } else {
                        self.symbol( SyntaxKind::LT, "<", tokens);
                    }
                }
                '.' => {
                    if let Some(c) = self.peek() {
                        // If the . is part of a number we push it
                        if c.is_numeric() {
                            self.working.push('.')
                        } else {
                            self.symbol( SyntaxKind::DOT, ".", tokens);
                        }
                    } else {
                        self.symbol( SyntaxKind::DOT, ".", tokens)
                    }
                }
                '/' => {
                    if let Some('/') = self.peek() {
                        self.next();
                        let mut peek = self.peek();
                        let mut comment = String::new();
                        while peek != Some('\n') && peek != None {
                            comment.push(self.next().unwrap());
                            peek = self.peek();
                        }
                        self.symbol( SyntaxKind::COMMENT, "//", tokens);
                    } else {
                        self.symbol( SyntaxKind::SLASH, "/", tokens);
                    }
                }
                ' ' => {
                    let mut length = 1;
                    let mut col = self.col;
                    while let Some(' ') = self.peek() {
                         length += 1;
                         self.next();
                    }
                    if let Some(tk) = self.scan_working() {
                        // Update the column as we may have scanned a new token
                        col = *tk.col.end() + 1;
                        tokens.push(tk);
                    }
                    tokens.push(Token::new(SyntaxKind::WHITESPACE, col..=(col + length - 1)));
                    self.col += length;
                }
                '+' => self.symbol( SyntaxKind::PLUS, "+", tokens),
                '-' => self.symbol( SyntaxKind::MINUS, "-", tokens),
                '*' => self.symbol( SyntaxKind::STAR, "*", tokens),
                '(' => self.symbol( SyntaxKind::L_PAREN, "(", tokens),
                ')' => self.symbol( SyntaxKind::R_PAREN, ")", tokens),
                '{' => self.symbol(SyntaxKind::L_BRACE, "{", tokens),
                '}' => self.symbol(SyntaxKind::R_BRACE, "}", tokens),
                '!' => self.symbol( SyntaxKind::EXCLAMATION, "!", tokens),
                ',' => self.symbol(SyntaxKind::COMMA, ",", tokens),
                ':' => self.symbol(SyntaxKind::COLON, ":", tokens),
                '\n' => self.symbol( SyntaxKind::NL, "\n", tokens),
                c => {
                    if c.is_alphanumeric() {
                        self.working.push(c)
                    } else {
                        self.symbol( SyntaxKind::ERROR, &c.to_string(), tokens)
                    }
                }
            }
            self.scan(tokens);
        } else {
            if let Some(tk) = self.scan_working() {
                tokens.push(tk);
            }
        }
    }
}

#[cfg(test)]
mod lexer_test {
    use super::Token;

    fn test_base(input: &str) -> String {
        let lexer = super::Lexer::new(input);
        let tokens = lexer.start_scan();
        tokens.into_iter().map(|t| {
            input.get(*t.col.start() as usize..=*t.col.end() as usize).unwrap()
        }).collect::<String>()
    }

    #[test]
    fn symbol_kinds() {
        let output = test_base("& | + - * / > < = ( ) ! . \n == >= <= && || { }");
        insta::assert_snapshot!(output);
    }

    #[test]
    fn keywords_and_ident() {
        let input = "and sometoiurng or skngeing";
        let lexer = super::Lexer::new(input);
        let tokens = lexer.start_scan();
        let output = tokens
            .into_iter()
            .map(|p| {
                match p {
                Token {ty: super::SyntaxKind::KW_AND, ..} => format!("|and:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::KW_OR, ..} => format!("|or:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::WHITESPACE, ..} => format!("|whitespace:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::IDENT, ..} => format!("|ident:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::SOURCE_FILE, ..}  => format!("|src:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                _ => format!("error:{}", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
            }})
            .collect::<String>();
        insta::assert_snapshot!(output);
    }

    #[test]
    fn numbers() {
        let input = "1.23 42";
        let lexer = super::Lexer::new(input);
        let tokens = lexer.start_scan();
        let output = tokens
            .into_iter()
            .map(|p| match p {
                Token {ty: super::SyntaxKind::FLOAT, .. } => format!("|float:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::INT, .. } => format!("|int:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::WHITESPACE, .. } => format!("|whitespace:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                Token {ty: super::SyntaxKind::SOURCE_FILE, .. }  => format!("|src:{}|", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
                _ => format!("error:{}", input.get(*p.col.start() as usize..=*p.col.end() as usize).unwrap()),
            })
            .collect::<String>();
        insta::assert_snapshot!(output);
    }
}
