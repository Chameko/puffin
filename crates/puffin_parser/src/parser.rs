use std::{iter::Peekable, vec::IntoIter};
use puffin_error::{Level, compiler::{CompilerError, Output, Highlight, CompilerErrorType} };
use puffin_ast::SyntaxKind;
use crate::lexer::Token;
use rowan::{GreenNode, GreenNodeBuilder};

pub struct Parse<'a> {
    pub green_node: GreenNode,
    pub errors: Vec<CompilerError<'a>>,
}

/// Parses a token stream (Vec<Token>) into a [Parse] result
pub struct Parser<'a> {
    /// The token stream for the parser
    tokens: Peekable<IntoIter<Token>>,
    /// The builder used to build the concrete syntax tree
    builder: GreenNodeBuilder<'static>,
    /// The errors the parser accumilates
    errors: Vec<CompilerError<'a>>,
    /// The source of the parser with the file name and the lines
    src: (String, &'a Vec<&'a str>),
}

impl<'a> Parser<'a> {
    /// Create a new parser
    pub fn new(tokens: Vec<Token>, file: &str, lines: &'a Vec<&'a str>) -> Self {
        Self {
            tokens: tokens.into_iter().peekable(),
            builder: GreenNodeBuilder::new(),
            errors: vec![],
            src: (file.to_string(), lines),
        }
    }

    pub fn parse(mut self) -> Parse<'a> {
        self.builder.start_node(SyntaxKind::SOURCE_FILE.into());
        self.expr(0);
        self.builder.finish_node();
        let green_node = self.builder.finish();
        Parse {
            green_node,
            errors: self.errors
        }
    }

    /// Parses an expression
    fn expr(&mut self, min_bp: u8) {
        self.skip_whitespace();
        let checkpoint = self.builder.checkpoint();

        match self.tokens.next() {
            // Literals
            Some( lhs@Token { ty: SyntaxKind::INT | SyntaxKind::FLOAT | SyntaxKind::IDENT, ..}) => {
                self.builder.token(lhs.ty.into(), &lhs.get_text(self.src.1));
            },
            // Grouping
            Some( Token { ty: SyntaxKind::R_PAREN, ..}) => {}
            Some(tk@Token{ ty: SyntaxKind::L_PAREN, .. }) => {
                self.builder.start_node(SyntaxKind::PAREN_EXPR.into());
                // Consume the left paren `(`
                self.builder.token(tk.ty.into(), &tk.get_text(self.src.1));
                // Parse the internal expression
                self.expr(0);
                // End with `)` in peek
                if let Some(tk@Token {ty: SyntaxKind::R_PAREN, ..}) = self.tokens.peek() {
                    self.builder.token(SyntaxKind::R_PAREN.into(), tk.get_text(self.src.1));
                    self.tokens.next();
                } else {
                    // Report error if missing `)`
                    self.errors.push(generic_error(
                        &tk,
                        &self,
                        CompilerErrorType::ExpectedSymbol,
                        "expected )",
                        &format!("expected ) found {}", tk.get_text(self.src.1))
                    ));
                }
                self.builder.finish_node();
            }
            // Prefix
            Some(tk) => {
                if let Some(((), rbp)) = Self::prefix_binding_power(&tk.ty) {
                    self.builder.start_node(SyntaxKind::PREFIX_EXPR.into());
                    self.builder.token(tk.ty.into(), &tk.get_text(self.src.1));
                    self.expr(rbp);
                    self.builder.finish_node();
                } else {
                    self.errors.push(generic_error(
                        &tk,
                        &self,
                        CompilerErrorType::UnexpectedSymbol,
                        "unexpected symbol",
                        &format!("unexpected symbol {}", tk.get_text(self.src.1))
                    ));
                    self.builder.token(SyntaxKind::ERROR.into(), &tk.get_text(self.src.1))
                }
            }
            // EOF
            None => return,
        };


        loop {
            self.skip_whitespace();

            let op = if let Some(tk) = self.tokens.peek() {
                tk
            } else {
                return;
            };

            if let Some((lbp, rbp)) = Self::infix_binding_power(&op.ty) {
                if lbp < min_bp {
                    return;
                }

                self.builder.start_node_at(checkpoint, SyntaxKind::BIN_EXPR.into());
                self.builder.token(op.ty.into(), &op.get_text(self.src.1));
                self.tokens.next();
                self.expr(rbp);
                self.builder.finish_node();
            } else {
                if op.ty == SyntaxKind::R_PAREN {
                    return;
                }
                let text = op.get_text(self.src.1);
                let hl = Highlight::new(op.line, op.col.clone(), "unexpected symbol", Level::Error);
                let output = vec![
                    Output::Msg(format!("unexpected symbol {}", text)),
                    Output::Code{ highlight: vec![hl], lines: vec![(op.line, self.src.1[op.line - 1])], src: self.src.0.clone() }
                ];
                let error = CompilerError::new(CompilerErrorType::UnexpectedSymbol, Level::Error, output);
                self.errors.push(error);
                self.builder.token(SyntaxKind::ERROR.into(), text);
                return;
            }
        }
    }

    /// Skips whitespace and adds it to the tree
    fn skip_whitespace(&mut self) {
        while let Some(Token { ty: SyntaxKind::WHITESPACE, .. }) = self.tokens.peek() {
            let tk = self.tokens.next().expect("Expected a token");
            self.builder.token(tk.ty.into(), &tk.get_text(self.src.1));
        }
    }

    /// Provides the binding power of the symbol in an infix position (the strength of its hold on its surrounding compenents)
    const fn infix_binding_power(op: &SyntaxKind) -> Option<(u8, u8)> {
        use SyntaxKind::*;
        match op {
            PLUS | MINUS => Some((1, 2)),
            STAR | SLASH => Some((3, 4)),
            _ => None,
        }
    }

    /// Provides the binding power of the symbol in a prefix position
    const fn prefix_binding_power(op: &SyntaxKind) -> Option<((), u8)> {
        use SyntaxKind::*;
        match op {
           MINUS => Some(((), 5)),
            _ => None,
        }
    }
}

/// Creates a compiler error where the line the supplied [`Token`] is on is outputed with that same [`Token`] highlighted
fn generic_error<'a>(tk: &Token, parser: &Parser<'a>, ty: CompilerErrorType, hl_msg: &str, output_msg: &str) -> CompilerError<'a> {
    let hl = Highlight::new(tk.line, tk.col.clone(), hl_msg, Level::Error);
    let output = vec![
        Output::Msg(output_msg.to_string()),
        Output::Code{ highlight: vec![hl], lines: vec![(tk.line, &parser.src.1[tk.line - 1])], src: parser.src.0.clone() }
    ];
    CompilerError::new(ty, Level::Error, output)
}

/// Helper function to output the CST in a readable manner
pub fn output_cst(cst: &rowan::GreenNodeData, mut output: String, offset: &mut u32, indent: usize) -> String {
    let len = cst.text_len();
    output.push_str( &format!("{}{}@{}..{}\n", " ".repeat(indent * 4), SyntaxKind::from(cst.kind().0), offset, *offset + u32::from(len)));
    for child in cst.children() {
        match child {
            rowan::NodeOrToken::Node(n) => { output = output_cst(n, output, offset, indent + 1); },
            rowan::NodeOrToken::Token(t) => {
                let len = t.text_len();
                output.push_str( &format!("{}{} |{:?}@{}..{}\n", " ".repeat((indent + 1) * 4), t.text(),SyntaxKind::from(t.kind().0), offset, *offset + u32::from(len)));
                *offset += u32::from(len);
            },
        }
    }
    output
}

#[cfg(test)]
mod parser_tests {
    use crate::lexer::{Lexer, Token};

    use super::Parser;
    use super::output_cst;

    fn scan_tokens(src: &str) -> Vec<Token> {
        let lexer = Lexer::new(src);
        lexer.start_scan()
    }

    #[test]
    fn just_number() {
        let src = vec!["1"];
        let parser = Parser::new(scan_tokens("1"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn simple_expr() {
        let src = vec!["1 + 2"];
        let parser = Parser::new(scan_tokens("1 + 2"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn multiple_simple() {
        let src = vec!["1 + 2 + 3 + 4"];
        let parser = Parser::new(scan_tokens("1 + 2 + 3 + 4"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn prefix_operation() {
        let src = vec!["-1 + 2"];
        let parser = Parser::new(scan_tokens("-1 + 2"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn multiple_prefix_operation() {
        let src = vec!["--1 + -2"];
        let parser = Parser::new(scan_tokens("--1 + -2"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn order_of_operations() {
        let src = vec!["1 + 2 * 5 - 3"];
        let parser = Parser::new(scan_tokens("1 + 2 * 5 - 3"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn paren() {
        let src = vec!["1 + 2 * (3 - 2) + (1 * 2)"];
        let parser = Parser::new(scan_tokens("1 + 2 * (3 - 2) + (1 * 2)"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn paren_2() {
        let src = vec!["(1 + 2 + 3)"];
        let parser = Parser::new(scan_tokens("(1 + 2 + 3)"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }

    #[test]
    fn paren_3() {
        let src = vec!["(1 + (1)) - (2 + 1 + 2)"];
        let parser = Parser::new(scan_tokens("(1 + (1)) - (2 + 1 + 2)"), "test.pf", &src);
        let parse = parser.parse();
        assert_eq!(parse.errors.len(), 0);
        let mut offset = 0;
        let output = output_cst(&parse.green_node, String::new(), &mut offset, 0);
        insta::assert_yaml_snapshot!(output);
    }
}
