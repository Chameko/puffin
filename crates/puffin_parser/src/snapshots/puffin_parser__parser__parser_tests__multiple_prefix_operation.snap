---
source: crates/puffin_parser/src/parser.rs
expression: output
---
SOURCE_FILE@0..8
    EXPR_STMT@0..8
        BIN_EXPR@0..8
            PREFIX_EXPR@0..4
                - |MINUS@0..1
                PREFIX_EXPR@1..4
                    - |MINUS@1..2
                    1 |INT@2..3
                      |WHITESPACE@3..4
            + |PLUS@4..5
              |WHITESPACE@5..6
            PREFIX_EXPR@6..8
                - |MINUS@6..7
                2 |INT@7..8

